File: apps/server/src/watchers/logs.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import chokidar, { type FSWatcher } from 'chokidar';
import type { FastifyBaseLogger } from 'fastify';
import { parseLogLine } from '@tarkov-checker/shared';
import type { Hub } from '../sse.js';

export interface LogsWatcher {
  stop: () => Promise<void>;
}

const SESSION_FOLDER_PREFIX = 'log_';
const APPLICATION_LOG_RE = /^.+ application_(\d+)\.log$/;

/**
 * Tail follower for a single growing log file. Tracks byte offset and
 * line-buffers partial reads (Tarkov writes lines with a trailing newline,
 * but a `change` event can fire mid-line). Reset on truncation (offset >
 * size) — rare for Tarkov logs but cheap to handle.
 */
class LogTail {
  private offset = 0;
  private partial = '';
  private readonly watcher: FSWatcher;
  private reading = false;
  private rereadPending = false;
  private stopped = false;

  constructor(
    private readonly filePath: string,
    private readonly onLine: (line: string) => void,
    private readonly log: FastifyBaseLogger,
  ) {
    try {
      // Start at EOF — we don't want to replay historical map switches on
      // startup. (Initial-state seeding is done explicitly by the parent
      // watcher before this tail is created.)
      this.offset = fs.statSync(filePath).size;
    } catch (err) {
      this.log.warn({ err, filePath }, 'log tail: initial stat failed');
    }

    this.watcher = chokidar.watch(filePath, {
      ignoreInitial: true,
      persistent: true,
      // Tail-style follow needs polling on Windows — fs.watch fires once per
      // OS notification, which Tarkov's flush cadence (~1Hz) tends to miss.
      usePolling: true,
      interval: 300,
    });
    this.watcher.on('change', () => void this.readNew());
    this.watcher.on('error', (err: unknown) => {
      this.log.warn({ err, filePath }, 'log tail: chokidar error');
    });
  }

  private async readNew(): Promise<void> {
    if (this.stopped) return;
    if (this.reading) {
      // Coalesce: a second change while we're mid-read just sets a flag.
      this.rereadPending = true;
      return;
    }
    this.reading = true;
    try {
      const stat = await fs.promises.stat(this.filePath);
      if (stat.size < this.offset) {
        // Truncation or rotation — start over from the top.
        this.offset = 0;
        this.partial = '';
      }
      if (stat.size === this.offset) return;
      const chunks: Buffer[] = [];
      await new Promise<void>((resolve, reject) => {
        const stream = fs.createReadStream(this.filePath, {
          start: this.offset,
          end: stat.size - 1,
        });
        stream.on('data', (c) => chunks.push(c as Buffer));
        stream.on('end', () => resolve());
        stream.on('error', reject);
      });
      this.offset = stat.size;
      const text = this.partial + Buffer.concat(chunks).toString('utf-8');
      const lines = text.split(/\r?\n/);
      this.partial = lines.pop() ?? '';
      for (const line of lines) {
        if (line.length === 0) continue;
        this.onLine(line);
      }
    } catch (err) {
      this.log.warn({ err, filePath: this.filePath }, 'log tail: read failed');
    } finally {
      this.reading = false;
      if (this.rereadPending) {
        this.rereadPending = false;
        void this.readNew();
      }
    }
  }

  async stop(): Promise<void> {
    this.stopped = true;
    await this.watcher.close();
  }
}

function listSessionFolders(logsDir: string): string[] {
  try {
    return fs
      .readdirSync(logsDir, { withFileTypes: true })
      .filter((e) => e.isDirectory() && e.name.startsWith(SESSION_FOLDER_PREFIX))
      .map((e) => e.name)
      .sort();
  } catch {
    return [];
  }
}

function pickLatestApplicationLog(sessionFolder: string): string | null {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(sessionFolder, { withFileTypes: true });
  } catch {
    return null;
  }
  let best: { file: string; suffix: number } | null = null;
  for (const e of entries) {
    if (!e.isFile()) continue;
    const m = APPLICATION_LOG_RE.exec(e.name);
    if (!m) continue;
    const suffix = Number(m[1]);
    if (!best || suffix > best.suffix) best = { file: e.name, suffix };
  }
  return best ? path.join(sessionFolder, best.file) : null;
}

/**
 * Scans an existing application log from the end for the most recent line
 * that yields a mapId. Lets the overlay snap to the correct map on startup
 * when Tarkov is already mid-raid (or in the lobby after a recent extract).
 *
 * Reads the tail of the file (~64 KB) — Tarkov maps fire several lines per
 * load, so a small window is enough; we don't need to walk gigabytes.
 */
function findLatestMapIdInFile(filePath: string): string | null {
  let fd: number | null = null;
  try {
    fd = fs.openSync(filePath, 'r');
    const stat = fs.fstatSync(fd);
    const windowSize = Math.min(stat.size, 64 * 1024);
    if (windowSize === 0) return null;
    const buf = Buffer.alloc(windowSize);
    fs.readSync(fd, buf, 0, windowSize, stat.size - windowSize);
    const lines = buf.toString('utf-8').split(/\r?\n/);
    for (let i = lines.length - 1; i >= 0; i--) {
      const parsed = parseLogLine(lines[i]!);
      if (parsed) return parsed.rawMapId;
    }
    return null;
  } catch {
    return null;
  } finally {
    if (fd !== null) {
      try {
        fs.closeSync(fd);
      } catch {
        /* ignore */
      }
    }
  }
}

/**
 * Watches Tarkov's logs directory and emits `map-change` events whenever the
 * game writes a recognised map-load line to the active session's application
 * log. Lifecycle:
 *
 *   1. On start: find the latest `log_*` folder + its highest
 *      `application_NNN.log`, scan its tail for a recent map id, emit one
 *      `map-change` if found (overlay snaps to current/last map).
 *   2. Tail that file from EOF.
 *   3. Watch `logsDir` for new `log_*` folders (new game launch). When one
 *      appears, drop the previous tail and re-seed from the new folder.
 *   4. Watch the active folder for higher-suffix `application_NNN.log` files
 *      (within-session rotation) and switch the tail to the newer file.
 *
 * Dedupes consecutive emits of the same id so the explicit `rcid:` line and
 * the later `Location:` confirmation collapse into a single broadcast.
 */
export function startLogsWatcher(
  logsDir: string,
  hub: Hub,
  log: FastifyBaseLogger,
): LogsWatcher {
  let lastEmittedMapId: string | null = null;
  let activeSession: { folder: string; logFile: string; tail: LogTail } | null = null;
  let sessionFolderWatcher: FSWatcher | null = null;
  let activeFolderWatcher: FSWatcher | null = null;
  let stopped = false;

  const emitMapChange = (rawMapId: string): void => {
    if (rawMapId === lastEmittedMapId) return;
    lastEmittedMapId = rawMapId;
    log.info({ rawMapId }, 'map change detected');
    hub.broadcast({ type: 'map-change', t: Date.now(), rawMapId });
  };

  const handleLine = (line: string): void => {
    const parsed = parseLogLine(line);
    if (parsed) emitMapChange(parsed.rawMapId);
  };

  const closeActiveFolderWatcher = async (): Promise<void> => {
    if (activeFolderWatcher) {
      await activeFolderWatcher.close();
      activeFolderWatcher = null;
    }
  };

  const switchToLogFile = async (folder: string, logFile: string): Promise<void> => {
    if (activeSession?.logFile === logFile) return;
    if (activeSession) {
      await activeSession.tail.stop();
    }
    activeSession = { folder, logFile, tail: new LogTail(logFile, handleLine, log) };
    log.info({ folder, logFile }, 'tailing log file');
  };

  const attachToSession = async (folder: string, seedFromExisting: boolean): Promise<void> => {
    await closeActiveFolderWatcher();
    if (activeSession) {
      await activeSession.tail.stop();
      activeSession = null;
    }
    const logFile = pickLatestApplicationLog(folder);
    if (!logFile) {
      // Folder just appeared but the game hasn't written any log file yet —
      // wait for it via the per-folder watcher below.
      log.info({ folder }, 'log session folder has no application_NNN.log yet');
    } else {
      if (seedFromExisting) {
        const seeded = findLatestMapIdInFile(logFile);
        if (seeded) {
          log.info({ rawMapId: seeded }, 'seeding map from latest log entry');
          emitMapChange(seeded);
        }
      }
      await switchToLogFile(folder, logFile);
    }

    // Watch the session folder for rotations (new `application_NNN.log` with
    // a higher suffix, or the very first one if the folder was empty on
    // arrival).
    activeFolderWatcher = chokidar.watch(folder, {
      ignoreInitial: true,
      persistent: true,
      depth: 0,
    });
    activeFolderWatcher.on('add', (filePath: string) => {
      const m = APPLICATION_LOG_RE.exec(path.basename(filePath));
      if (!m) return;
      const newSuffix = Number(m[1]);
      const currentMatch = activeSession
        ? APPLICATION_LOG_RE.exec(path.basename(activeSession.logFile))
        : null;
      const currentSuffix = currentMatch ? Number(currentMatch[1]) : -1;
      if (newSuffix > currentSuffix) {
        void switchToLogFile(folder, filePath);
      }
    });
    activeFolderWatcher.on('error', (err: unknown) => {
      log.warn({ err, folder }, 'log folder watcher error');
    });
  };

  const onSessionFolderSeen = (dir: string): void => {
    if (stopped) return;
    const name = path.basename(dir);
    if (!name.startsWith(SESSION_FOLDER_PREFIX)) return;
    if (activeSession && name <= path.basename(activeSession.folder)) return;
    log.info({ folder: dir }, 'new tarkov log session detected');
    // Don't seed from a brand-new session: it has nothing in it yet, and if
    // it ever does, the tail will pick those lines up live.
    void attachToSession(dir, /* seedFromExisting */ false);
  };

  // Initial scan: pick the latest existing log folder and seed from it.
  const folders = listSessionFolders(logsDir);
  if (folders.length > 0) {
    const latest = folders[folders.length - 1]!;
    void attachToSession(path.join(logsDir, latest), /* seedFromExisting */ true);
  } else {
    log.info({ logsDir }, 'no existing tarkov log sessions yet — waiting for first launch');
  }

  // Watch the parent for fresh session folders (next game launch).
  sessionFolderWatcher = chokidar.watch(logsDir, {
    ignoreInitial: true,
    persistent: true,
    depth: 0,
  });
  sessionFolderWatcher.on('addDir', onSessionFolderSeen);
  sessionFolderWatcher.on('error', (err: unknown) => {
    log.warn({ err, logsDir }, 'logs dir watcher error');
  });

  log.info({ logsDir }, 'logs watcher started');

  return {
    stop: async () => {
      stopped = true;
      await Promise.all([
        sessionFolderWatcher?.close(),
        closeActiveFolderWatcher(),
        activeSession?.tail.stop(),
      ]);
      sessionFolderWatcher = null;
      activeSession = null;
    },
  };
}

File: apps/desktop/src-tauri/src/server/logs.rs
//! Port of `apps/server/src/watchers/logs.ts` + the parser helper from
//! `packages/shared/src/parse-log.ts`. Parser regex is mirrored literally —
//! both ports must extract the same `rawMapId` from any given line.

use anyhow::{Context, Result};
use notify::{RecursiveMode, Watcher};
use notify_debouncer_full::{new_debouncer, Debouncer, FileIdMap};
use once_cell::sync::Lazy;
use regex::Regex;
use serde::Serialize;
use std::fs::{self, File};
use std::io::{Read, Seek, SeekFrom};
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{mpsc, Arc, Mutex};
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tauri::{AppHandle, Emitter};

const SESSION_FOLDER_PREFIX: &str = "log_";

static SCENE_PRESET_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\brcid:([A-Za-z0-9_]+)\.scenespreset\.asset\b").expect("static regex"));
static TRANSIT_LOCATION_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\[Transit\].*\bLocations:([A-Za-z0-9_]+)").expect("static regex"));
static TRACE_LOCATION_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\bLocation:\s+([A-Za-z0-9_]+)\s*,\s*Sid:").expect("static regex"));
static APPLICATION_LOG_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^.+ application_(\d+)\.log$").expect("static regex"));

/// Pure parser — direct port of `parseLogLine` in TS. See that doc-comment
/// for the format reference. Result is lowercased so callers can use it as
/// a stable lookup key against the canonical map alphabet.
pub fn parse_log_line(line: &str) -> Option<String> {
    let m = SCENE_PRESET_RE
        .captures(line)
        .or_else(|| TRANSIT_LOCATION_RE.captures(line))
        .or_else(|| TRACE_LOCATION_RE.captures(line))?;
    Some(m.get(1)?.as_str().to_ascii_lowercase())
}

#[derive(Debug, Clone, Serialize)]
struct MapChangePayload {
    t: i64,
    #[serde(rename = "rawMapId")]
    raw_map_id: String,
}

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// State machine kept inside the Mutex. `offset` is `None` until a log file
/// is actually open; the folder may exist before any application_NNN.log is
/// written to it (very early in a game launch).
struct State {
    last_emitted_map_id: Option<String>,
    active_folder: Option<PathBuf>,
    active_log: Option<ActiveLog>,
}

struct ActiveLog {
    path: PathBuf,
    suffix: u32,
    offset: u64,
    partial: String,
}

fn list_session_folders(logs_dir: &Path) -> Vec<PathBuf> {
    let mut out: Vec<PathBuf> = match fs::read_dir(logs_dir) {
        Ok(it) => it
            .filter_map(|e| e.ok())
            .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
            .map(|e| e.path())
            .filter(|p| {
                p.file_name()
                    .and_then(|n| n.to_str())
                    .map(|s| s.starts_with(SESSION_FOLDER_PREFIX))
                    .unwrap_or(false)
            })
            .collect(),
        Err(_) => Vec::new(),
    };
    out.sort();
    out
}

fn pick_latest_application_log(folder: &Path) -> Option<(PathBuf, u32)> {
    let entries = fs::read_dir(folder).ok()?;
    let mut best: Option<(PathBuf, u32)> = None;
    for entry in entries.flatten() {
        if !entry.file_type().map(|t| t.is_file()).unwrap_or(false) {
            continue;
        }
        let name = entry.file_name();
        let Some(name_str) = name.to_str() else { continue };
        let Some(caps) = APPLICATION_LOG_RE.captures(name_str) else { continue };
        let Some(suffix_str) = caps.get(1).map(|m| m.as_str()) else { continue };
        let Ok(suffix) = suffix_str.parse::<u32>() else { continue };
        if best.as_ref().map_or(true, |(_, s)| suffix > *s) {
            best = Some((entry.path(), suffix));
        }
    }
    best
}

/// Reads up to the last `WINDOW` bytes of the file and scans backwards for
/// the most recent line that yields a mapId. Used once on startup so the
/// overlay snaps to the right map when Tarkov is already mid-raid.
fn find_latest_map_id_in_file(path: &Path) -> Option<String> {
    const WINDOW: u64 = 64 * 1024;
    let mut file = File::open(path).ok()?;
    let size = file.metadata().ok()?.len();
    if size == 0 {
        return None;
    }
    let read_from = size.saturating_sub(WINDOW);
    file.seek(SeekFrom::Start(read_from)).ok()?;
    let mut buf = Vec::with_capacity((size - read_from) as usize);
    file.read_to_end(&mut buf).ok()?;
    let text = String::from_utf8_lossy(&buf);
    for line in text.lines().rev() {
        if let Some(id) = parse_log_line(line) {
            return Some(id);
        }
    }
    None
}

fn emit_map_change(app: &AppHandle, state: &Mutex<State>, raw_map_id: String) {
    let mut guard = state.lock().expect("logs state poisoned");
    if guard
        .last_emitted_map_id
        .as_deref()
        .is_some_and(|prev| prev == raw_map_id)
    {
        return;
    }
    guard.last_emitted_map_id = Some(raw_map_id.clone());
    drop(guard);

    let payload = MapChangePayload {
        t: now_ms(),
        raw_map_id,
    };
    if let Err(err) = app.emit("map-change", &payload) {
        eprintln!("[logs-watcher] emit failed: {err}");
    }
}

/// Seek + read appended bytes for the currently-tailed log file, parse line
/// by line, emit on hit. Idempotent — safe to call on every poll tick.
fn read_appended(app: &AppHandle, state: &Mutex<State>) {
    // Snapshot the path + offset under the lock so the actual I/O happens
    // outside it.
    let snapshot: Option<(PathBuf, u64, String)> = {
        let guard = state.lock().expect("logs state poisoned");
        guard
            .active_log
            .as_ref()
            .map(|a| (a.path.clone(), a.offset, a.partial.clone()))
    };
    let Some((path, mut offset, mut partial)) = snapshot else {
        return;
    };

    let Ok(metadata) = fs::metadata(&path) else { return };
    let size = metadata.len();
    if size < offset {
        // Truncation / rotation — restart from the top.
        offset = 0;
        partial.clear();
    }
    if size == offset {
        return;
    }

    let Ok(mut file) = File::open(&path) else { return };
    if file.seek(SeekFrom::Start(offset)).is_err() {
        return;
    }
    let to_read = (size - offset) as usize;
    let mut buf = vec![0u8; to_read];
    let read_bytes = match file.read(&mut buf) {
        Ok(n) => n,
        Err(_) => return,
    };
    buf.truncate(read_bytes);
    offset += read_bytes as u64;

    partial.push_str(&String::from_utf8_lossy(&buf));
    let mut lines: Vec<&str> = partial.split('\n').collect();
    // The trailing piece is whatever follows the last '\n' — keep it for
    // the next tick.
    let leftover = lines.pop().unwrap_or("").to_owned();
    let mut hits: Vec<String> = Vec::new();
    for line in lines {
        let line = line.strip_suffix('\r').unwrap_or(line);
        if let Some(id) = parse_log_line(line) {
            hits.push(id);
        }
    }

    {
        let mut guard = state.lock().expect("logs state poisoned");
        if let Some(active) = guard.active_log.as_mut() {
            // Only persist if the file we just read is still the active one
            // (a folder-watcher event could have swapped it underneath us).
            if active.path == path {
                active.offset = offset;
                active.partial = leftover;
            }
        }
    }

    for id in hits {
        emit_map_change(app, state, id);
    }
}

/// Switch the tail to a new `log_*` session folder (and its highest
/// application_NNN.log if one already exists). `seed_from_existing` is true
/// only on startup — when we discover a brand-new folder mid-session, the
/// folder is empty and we just wait for the tail to pick up new lines.
fn attach_to_session(
    app: &AppHandle,
    state: &Mutex<State>,
    folder: PathBuf,
    seed_from_existing: bool,
) {
    let picked = pick_latest_application_log(&folder);

    let seed_id = if seed_from_existing {
        picked
            .as_ref()
            .and_then(|(path, _)| find_latest_map_id_in_file(path))
    } else {
        None
    };

    {
        let mut guard = state.lock().expect("logs state poisoned");
        guard.active_folder = Some(folder);
        guard.active_log = picked.map(|(path, suffix)| {
            let offset = fs::metadata(&path).map(|m| m.len()).unwrap_or(0);
            ActiveLog {
                path,
                suffix,
                offset,
                partial: String::new(),
            }
        });
    }

    if let Some(id) = seed_id {
        emit_map_change(app, state, id);
    }
}

/// Check the active folder for a higher-suffix application_NNN.log file and
/// switch to it if present. Cheap — a single read_dir per poll tick.
fn check_within_session_rotation(state: &Mutex<State>) {
    let snapshot = {
        let guard = state.lock().expect("logs state poisoned");
        guard.active_folder.clone().map(|folder| {
            let current_suffix = guard.active_log.as_ref().map(|a| a.suffix);
            (folder, current_suffix)
        })
    };
    let Some((folder, current_suffix)) = snapshot else { return };
    let Some((new_path, new_suffix)) = pick_latest_application_log(&folder) else { return };
    if current_suffix.is_some_and(|s| new_suffix <= s) {
        return;
    }
    // Promote to the newer file. Start at its current EOF so we don't replay
    // historical bytes from a freshly-rotated _001 file (its content is
    // either empty or a continuation).
    let mut guard = state.lock().expect("logs state poisoned");
    if guard
        .active_log
        .as_ref()
        .is_some_and(|a| a.path == new_path)
    {
        return;
    }
    let offset = fs::metadata(&new_path).map(|m| m.len()).unwrap_or(0);
    guard.active_log = Some(ActiveLog {
        path: new_path,
        suffix: new_suffix,
        offset,
        partial: String::new(),
    });
}

/// Check the logs dir for a session folder newer than the active one. The
/// notify debouncer fires on folder creation, but the actual lookup happens
/// here so we don't have to thread name-comparison logic through the event
/// channel.
fn check_new_session_folder(app: &AppHandle, state: &Mutex<State>, logs_dir: &Path) {
    let folders = list_session_folders(logs_dir);
    let Some(latest) = folders.last().cloned() else { return };
    let current = {
        let guard = state.lock().expect("logs state poisoned");
        guard.active_folder.clone()
    };
    let is_newer = match &current {
        Some(c) => latest > *c,
        None => true,
    };
    if !is_newer {
        return;
    }
    eprintln!(
        "[logs-watcher] new tarkov log session detected: {}",
        latest.display()
    );
    // seed_from_existing=false: a fresh folder is empty; the tail will pick
    // up the next session's first map-load line live.
    attach_to_session(app, state, latest, false);
}

pub struct LogsWatcher {
    _debouncer: Debouncer<notify::RecommendedWatcher, FileIdMap>,
    stop_flag: Arc<AtomicBool>,
}

impl Drop for LogsWatcher {
    fn drop(&mut self) {
        self.stop_flag.store(true, Ordering::SeqCst);
    }
}

/// Start watching `logs_dir`. Lifecycle mirrors the Node port in
/// `apps/server/src/watchers/logs.ts` — see its doc-comment for the full
/// behaviour spec.
pub fn start(logs_dir: PathBuf, app: AppHandle) -> Result<LogsWatcher> {
    let state = Arc::new(Mutex::new(State {
        last_emitted_map_id: None,
        active_folder: None,
        active_log: None,
    }));

    // Initial seed from latest existing session, if any.
    if let Some(latest) = list_session_folders(&logs_dir).last().cloned() {
        attach_to_session(&app, &state, latest, true);
    } else {
        eprintln!(
            "[logs-watcher] no existing tarkov log sessions yet at {}",
            logs_dir.display()
        );
    }

    // Folder watcher — wakes up the poll thread quickly on new session
    // creation. The poll thread also checks periodically, so this is purely
    // a latency optimisation, not a correctness requirement.
    let (tx, rx) = mpsc::channel::<notify_debouncer_full::DebounceEventResult>();
    let mut debouncer = new_debouncer(Duration::from_millis(250), None, tx)
        .context("create fs debouncer for logs dir")?;
    debouncer
        .watcher()
        .watch(&logs_dir, RecursiveMode::NonRecursive)
        .with_context(|| format!("watch {}", logs_dir.display()))?;
    debouncer
        .cache()
        .add_root(&logs_dir, RecursiveMode::NonRecursive);

    let stop_flag = Arc::new(AtomicBool::new(false));

    // Folder event drain thread — its only job is to wake the poll thread
    // and let it observe the new folder. The actual switching logic lives
    // in the poll thread to keep state ownership single-threaded.
    let stop_for_folders = Arc::clone(&stop_flag);
    let app_for_folders = app.clone();
    let state_for_folders = Arc::clone(&state);
    let logs_dir_for_folders = logs_dir.clone();
    std::thread::Builder::new()
        .name("logs-folder-watcher".into())
        .spawn(move || {
            for _batch in rx {
                if stop_for_folders.load(Ordering::SeqCst) {
                    break;
                }
                check_new_session_folder(&app_for_folders, &state_for_folders, &logs_dir_for_folders);
            }
        })
        .context("spawn logs folder watcher thread")?;

    // Tail poll thread — 300 ms cadence matches the Node port's chokidar
    // polling interval. Also performs cheap within-session rotation
    // checks each tick (handles `_000` → `_001` rollover).
    let stop_for_tail = Arc::clone(&stop_flag);
    let app_for_tail = app.clone();
    let state_for_tail = Arc::clone(&state);
    std::thread::Builder::new()
        .name("logs-tail-watcher".into())
        .spawn(move || {
            while !stop_for_tail.load(Ordering::SeqCst) {
                std::thread::sleep(Duration::from_millis(300));
                check_within_session_rotation(&state_for_tail);
                read_appended(&app_for_tail, &state_for_tail);
            }
        })
        .context("spawn logs tail watcher thread")?;

    eprintln!("[logs-watcher] started on {}", logs_dir.display());

    Ok(LogsWatcher {
        _debouncer: debouncer,
        stop_flag,
    })
}
